File: ProblemSet/464 · 整数排序 II/python/464 · 整数排序 II 归并排序 递归实现.py
class Solution:
    """
    @param A: an integer array
    @return: nothing
    """
    def sortIntegers2(self, A):
        self.tmp = [None] * len(A)
        self.mergeSort(A, 0, len(A) - 1)

    def mergeSort(self, nums, left, right):
        if left < right:
            mid = left + ((right - left) >> 1)
            self.mergeSort(nums, left, mid)
            self.mergeSort(nums, mid + 1, right)
            self.merge(nums, left, mid, mid + 1, right)

    def merge(self, nums, lStart, lEnd, rStart, rEnd):
        i, j, pos = lStart, rStart, lStart
        while i <= lEnd and j <= rEnd:
            if nums[i] <= nums[j]:
                self.tmp[pos] = nums[i]
                i += 1
            else:
                self.tmp[pos] = nums[j]
                j += 1
            pos += 1
        self.tmp[pos : rEnd + 1] = nums[i : lEnd + 1] if i <= lEnd else nums[j : rEnd + 1]
        nums[lStart : rEnd + 1] = self.tmp[lStart : rEnd + 1]

File: ProblemSet/464 · 整数排序 II/java/464 · 整数排序 II 归并排序 递归实现.java
public class Solution {
    private int[] tmp;
    /**
     * @param A: an integer array
     * @return: nothing
     */
    public void sortIntegers2(int[] A) {
        tmp = new int[A.length];
        mergeSort(A, 0, A.length - 1);
    }

    private void mergeSort(int[] nums, int left, int right) {
        if (left < right) {
            int mid = left + ((right - left) >> 1);
            mergeSort(nums, left, mid);
            mergeSort(nums, mid + 1, right);
            merge(nums, left, mid, mid + 1, right);
        }
    }

    private void merge(int[] nums, int lStart, int lEnd, int rStart, int rEnd) {
        int i = lStart, j = rStart, pos = lStart;
        while (i <= lEnd && j <= rEnd) tmp[pos++] = nums[i] <= nums[j] ? nums[i++] : nums[j++];
        while (i <= lEnd) tmp[pos++] = nums[i++];
        while (j <= rEnd) tmp[pos++] = nums[j++];
        for (int k = lStart; k <= rEnd; ++k) nums[k] = tmp[k];
    }
}